tidy3d/plugins/adjoint/utils/penalty.py

```diff
@@ -134,3 +136,119 @@ def penalty_fn(radius):
 
         # return the average penalty over the points
         return jnp.sum(penalty_fn(rs)) / len(rs)
+
+
+class ErosionDilationPenalty(Penalty):
+    """Computes a penalty for erosion / dilation of a parameter map not being unity.
+    Accepts a parameter array normalized between 0 and 1.
+    Uses filtering and projection methods to erode and dilate the features within this array.
+    Measures the change in the array after eroding and dilating (and also dilating and eroding).
+    Returns a penalty proportional to the magnitude of this change.
+    The amount of change under dilation and erosion is minimized if the structure has large feature
+    sizes and large radius of curvature relative to the length scale.
+
+    Note
+    ----
+    For more details, refer to chapter 4 of Hammond, A., "High-Efficiency Topology Optimization
+    for Very Large-Scale Integrated-Photonics Inverse Design" (2022).
+    """
+
+    length_scale: pd.NonNegativeFloat = pd.Field(
+        ...,
+        title="Length Scale",
+        description="Length scale of erosion and dilation."
+        "Corresponds to ``radius`` in the :class:`ConicFilter` used for filtering."
+        "The parameter array is dilated and eroded by half of this value with each operation."
+        "Roughly corresponds to the desired minimum feature size and radius of curvature",
+        units=MICROMETER,
+    )
+
+    pixel_size: pd.PositiveFloat = pd.Field(
+        ...,
+        title="Pixel Size",
+        description="Size of each pixel in the array (must be the same along all dimensions)."
+        "Corresponds to ``design_region_dl`` in the :class:`ConicFilter` used for filtering.",
+        units=MICROMETER,
+    )
+
+    beta: pd.PositiveFloat = pd.Field(
+        1.0,
+        title="Projection Beta",
+        description="Strength of the ``tanh`` projection."
+        "Corresponds to ``beta`` in the :class:`BinaryProjector`."
+        "Higher values correspond to stronger discretization.",
+    )
+
+    penalty_strength = pd.Field(
+        1.0,
+        title="Penalty Strength",
+        description="Optional value to scale the returned penalty by.",
+    )
```

Comment:
I don't think we need a `penalty_strength` Field? It can be trivially applied after the penalty evaluation.

Reply:
yea I agree, wasn't sure so decided to put it in originally but will remove.